<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Martin Petr">

<title>slendr crash course</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="globe-slendr-2023-03-22_files/libs/clipboard/clipboard.min.js"></script>
<script src="globe-slendr-2023-03-22_files/libs/quarto-html/quarto.js"></script>
<script src="globe-slendr-2023-03-22_files/libs/quarto-html/popper.min.js"></script>
<script src="globe-slendr-2023-03-22_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="globe-slendr-2023-03-22_files/libs/quarto-html/anchor.min.js"></script>
<link href="globe-slendr-2023-03-22_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="globe-slendr-2023-03-22_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="globe-slendr-2023-03-22_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="globe-slendr-2023-03-22_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="globe-slendr-2023-03-22_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><em>slendr</em> crash course</h1>
<p class="subtitle lead">simulation tutorial at Globe</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Martin Petr </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">March 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="section" class="level1">
<h1></h1>
<div class="columns">
<div class="column" style="width:70%;">
<blockquote class="blockquote">
<p>Many problems in population genetics cannot be solved by a mathematician, no matter how gifted. [It is] clear that computer methods are very powerful. This is good. It [‚Ä¶] permits people with limited mathematical knowledge to work on important problems [‚Ä¶]</p>
</blockquote>
</div><div class="column" style="width:30%;">
<p><img src="images/crow.jpeg" class="img-fluid"></p>
<p><a href="https://en.wikipedia.org/wiki/James_F._Crow">James F. Crow</a> ‚Äì <a href="http://www.gnxp.com/blog/2006/06/10-questions-for-jim-crow.php">interview</a></p>
</div>
</div>
</section>
<section id="why-use-simulations" class="level1 page-columns page-full">
<h1>Why use simulations?</h1>
<ol type="1">
<li>Making sense of estimated statistics</li>
<li>Fitting model parameters</li>
<li>Ground truth for method work</li>
</ol>
<section id="making-sense-of-estimated-statistics" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="making-sense-of-estimated-statistics">Making sense of estimated statistics</h2>
<center>
<img src="images/elephants.png" class="img-fluid" style="width:65.0%">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p><a href="https://www.pnas.org/doi/10.1073/pnas.1720554115">Palkopoulou (2018)</a></p>
</div></div></section>
<section id="fitting-model-parameters-i.e.-abc" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="fitting-model-parameters-i.e.-abc">Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</h2>
<center>
<img src="images/abc_scheme.png" class="img-fluid" style="width:50.0%">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p>Image from <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">Wikipedia on ABC</a></p>
</div></div></section>
<section id="ground-truth-for-methods-work" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="ground-truth-for-methods-work">Ground truth for methods work</h2>
<center>
<img src="images/mcmc.png" class="img-fluid">
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p><a href="https://www.nature.com/articles/ng.3015">Schiffels and Durbin (2014)</a></p>
</div></div></section>
<section id="there-are-many-simulation-tools" class="level2">
<h2 class="anchored" data-anchor-id="there-are-many-simulation-tools">There are many simulation tools</h2>
<p><br></p>
<p>The most famous and widely used are <a href="https://messerlab.org/slim/">SLiM</a> and <a href="https://tskit.dev/msprime/docs/stable/intro.html"><em>msprime</em></a>.</p>
<div>
<p><br></p>
<p>Both are very powerful‚Ä¶</p>
<p>‚Ä¶ but require quite a bit of programming knowledge...</p>
<p>‚Ä¶ and a lot of code for non-trivial simulations (üêõü™≤üêú).</p>
</div>
<div class="fragment">
<p><br></p>
<center>
<strong>This is why we will focus on <a href="http://www.slendr.net"><em>slendr</em></a>,</strong><br>a new R interface for popgen modeling.
</center>
</div>
</section>
<section id="section-1" class="level2">
<h2 class="anchored" data-anchor-id="section-1"></h2>
<center>
<p><img src="images/slendr_logo.png" class="img-fluid" style="width:30.0%"></p>
<p><br></p>
<h2 class="anchored">
<a href="https://www.slendr.net">www.slendr.net</a>
</h2>
</center>
</section>
<section id="why-a-new-package-spatial-simulations" class="level2">
<h2 class="anchored" data-anchor-id="why-a-new-package-spatial-simulations">Why a new package? ‚Äì spatial simulations!</h2>
<div class="fragment">
<center>
<img src="images/animation.gif" class="img-fluid" style="width:70.0%">
</center>
</div>
</section>
<section id="why-a-new-package" class="level2">
<h2 class="anchored" data-anchor-id="why-a-new-package">Why a new package?</h2>
<div class="incremental">
<ul>
<li><p>Most researchers are not expert programmers</p></li>
<li><p>All but the most trivial simulations require lots of code</p></li>
</ul>
</div>
<div class="incremental">
<ul>
<li><p>90% <citation needed=""> of simulations are basically the same!</citation></p>
<ul>
<li><p>create populations (splits and <span class="math inline">\(N_e\)</span> changes)</p></li>
<li><p>specify gene flow (admixture rates and times)</p></li>
<li><p>save output (VCF, EIGENSTRAT)</p></li>
</ul></li>
</ul>
</div>
<div class="incremental">
<ul>
<li>Lot of code duplication across projects</li>
</ul>
</div>
</section>
</section>
<section id="motto-of-slendr" class="level1">
<h1>Motto of <em>slendr</em></h1>
<div>
<p><br></p>
<h3 class="anchored" data-anchor-id="motto-of-slendr">
‚ÄúSolve 90% of demographic modeling in popgen with 10% effort.‚Äù
</h3>
</div>
</section>
<section id="lets-get-started" class="level1 page-columns page-full">
<h1>Let‚Äôs get started</h1>
<section id="we-will-need-slendr-tidyverse" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="we-will-need-slendr-tidyverse">We will need <em>slendr</em> &amp; <em>tidyverse</em></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load Martin's favourite data analysis packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)     </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load slendr itself</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>=======================================================================
NOTE: Due to Python setup issues on some systems which have been
causing trouble particularly for novice users, calling library(slendr)
no longer activates slendr's Python environment automatically.

In order to use slendr's msprime back end or its tree-sequence
functionality, users must now activate slendr's Python environment
manually by executing init_env() after calling library(slendr).

(This note will be removed in the next major version of slendr.)
=======================================================================</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># setup slendr's internal machinery</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (you must run `setup_env()` first!)</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>You can safely ignore the message about missing SLiM.</p>
</div></div></section>
<section id="typical-steps-outline-of-this-tutorial" class="level2">
<h2 class="anchored" data-anchor-id="typical-steps-outline-of-this-tutorial">Typical steps (outline of this tutorial)</h2>
<p><br></p>
<ol type="1">
<li>create populations</li>
<li>schedule population splits</li>
<li>program <span class="math inline">\(N_e\)</span> size changes</li>
<li>specify gene-flow events</li>
<li>simulate sequence data</li>
<li>compute statistics</li>
</ol>
</section>
<section id="creating-a-population" class="level2">
<h2 class="anchored" data-anchor-id="creating-a-population">Creating a <code>population()</code></h2>
<p>Each needs a name, size and time of appearance (i.e., ‚Äúsplit‚Äù):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br></p>
<p>This creates a normal R object. Typing it in R gives us:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>pop1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop1 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 1: created as an ancestral population (N = 1000)</code></pre>
</div>
</div>
</div>
</section>
<section id="programming-population-splits" class="level2">
<h2 class="anchored" data-anchor-id="programming-population-splits">Programming population splits</h2>
<p>Splits are indicated by the <code>parent = &lt;pop&gt;</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br></p>
<p>The split is reported in the ‚Äúhistorical summary‚Äù:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>pop2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop2 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 50: split from pop1 (N = 100)</code></pre>
</div>
</div>
</div>
</section>
<section id="scheduling-resize-events-resize" class="level2">
<h2 class="anchored" data-anchor-id="scheduling-resize-events-resize">Scheduling resize events ‚Äì <code>resize()</code></h2>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb16-2"><a href="#cb16-2"></a>pop1_step <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop1, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>pop2_exp <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop2, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="tidyverse-style-pipe-interface" class="level2">
<h2 class="anchored" data-anchor-id="tidyverse-style-pipe-interface">Tidyverse-style <a href="https://magrittr.tidyverse.org">pipe</a> <code>%&gt;%</code> interface</h2>
<p>A more concise way to express the same thing.</p>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="a-more-complex-model" class="level2">
<h2 class="anchored" data-anchor-id="a-more-complex-model">A more complex model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">250</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">400</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">2400</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="last-step-before-simulation-compile_model" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="last-step-before-simulation-compile_model">Last step before simulation: <code>compile_model()</code></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Compilation takes a list of model components, performs internal consistency checks, returns a single model object.</strong>
</center>

<div class="no-row-height column-margin column-container"><div class="">
<p>The model is also compiled to disk which gives a nice additional layer of reproducibility. The exact location can be specified via <code>path =</code> argument to <code>compile_model()</code>.</p>
</div></div></section>
<section id="model-summary" class="level2">
<h2 class="anchored" data-anchor-id="model-summary">Model summary</h2>
<p>Again, typing the model into R prints a brief summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'model' object 
--------------------- 
populations: pop1, pop2, pop3, pop4, pop5 
geneflow events: [no geneflow]
generation time: 1 
time direction: forward 
total running length: 3000 model time units
model type: non-spatial

configuration files in: /private/var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T/RtmpUMgcTQ/file14f4e49b31aaa </code></pre>
</div>
</div>
</section>
<section id="model-visualization" class="level2">
<h2 class="anchored" data-anchor-id="model-visualization">Model visualization</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="globe-slendr-2023-03-22_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-1" class="level1 page-columns page-full">
<h1>Exercise #1</h1>
<section id="exercise-1-write-your-own-model" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="exercise-1-write-your-own-model">Exercise #1 ‚Äî write your own model!</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="fragment">
<p>You can use this ‚Äútemplate‚Äù:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">init_env</span>()</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>chimp <span class="ot">&lt;-</span> <span class="fu">population</span>(...)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co"># &lt;... rest of your code ...&gt;</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(chimp, ...),</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model) <span class="co"># verify visually</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div><div class="column" style="width:50%;">
<p><img src="images/intro_model1.png" class="img-fluid"></p>
</div>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>Don‚Äôt worry about gene flow just yet. We will add that at a later stage.</p>
<p>Feel free to include expansions and contractions (maybe in EUR at some point?).</p>
</div></div></section>
</section>
<section id="exercise-1-solution" class="level1">
<h1>Exercise #1 ‚Äî solution</h1>
<p>Solution is on GitHub at <a href="https://github.com/bodkan/globe-slendr-2023-03-22/tree/main/solutions/ex1.R">this</a> link.</p>
</section>
<section id="simulations" class="level1">
<h1>Simulations!</h1>
<section id="simulating-data" class="level2">
<h2 class="anchored" data-anchor-id="simulating-data">Simulating data</h2>
<p>We have a compiled <code>model</code> R object, how do we simulate data?</p>
<div class="incremental">
<p><em>slendr</em> has two built-in simulation engines:</p>
<ul>
<li>SLiM engine written <a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.slim">in Eidos/SLiM</a> (<code>slim()</code>)</li>
<li><em>msprime</em> engine written <a href="https://github.com/bodkan/slendr/blob/main/inst/scripts/script.py">in Python</a> (<code>msprime()</code>)</li>
</ul>
</div>
<div class="incremental">
<center>
<h4 class="anchored">
<strong>You don‚Äôt have to write any <em>msprime</em> or SLiM code!</strong>
</h4>
</center>
</div>
<div class="incremental">
<p><br> This is all that‚Äôs needed to simulate data from a <code>model</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>ts</code> is a so-called <strong>tree sequence</strong>.</p>
</div>
</section>
</section>
<section id="the-output-of-a-slendr-simulation-is-a-tree-sequence" class="level1">
<h1>The output of a <em>slendr</em> simulation is a <strong>tree sequence</strong></h1>
<section id="what-is-a-tree-sequence-see-tskit" class="level2">
<h2 class="anchored" data-anchor-id="what-is-a-tree-sequence-see-tskit">What is a tree sequence? (see <a href="https://tskit.dev">tskit</a>)</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/tree_sequence_diagram.png" class="img-fluid figure-img" style="width:80.0%"></p>
</figure>
</div>
<ul>
<li>a record of full genetic ancestry of a set of samples</li>
<li>an encoding of DNA sequence carried by those samples</li>
<li>an efficient analysis framework</li>
</ul>
</section>
</section>
<section id="why-tree-sequence" class="level1 page-columns page-full">
<h1>Why tree sequence?</h1>
<p><br></p>
<h3 class="anchored">
Why not VCF, EIGENSTRAT, etc.?
</h3>
<section id="what-we-usually-have" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-have">What we usually have</h2>
<center>
<img src="images/vcf_screenshot.png" class="img-fluid" style="width:90.0%">
</center>
</section>
<section id="what-we-usually-want" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-want">What we usually <em>want</em></h2>
<p>(As full as possible) a representation of our samples‚Äô history:</p>
<center>
<img src="images/tree_sequence_diagram.png" class="img-fluid">
</center>
<div class="fragment">
<center>
<strong>This is exactly what a tree sequence <em>is</em>.</strong>
</center>
</div>
</section>
<section id="mutation-can-be-done-after-simulation" class="level2">
<h2 class="anchored" data-anchor-id="mutation-can-be-done-after-simulation">Mutation can be done <em>after</em> simulation!</h2>
<p>If we have a simulated tree-sequence object <code>ts</code>, we can do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>ts_mutated <span class="ot">&lt;-</span> <span class="fu">ts_mutate</span>(ts, <span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br>Or, with a shortcut:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>We will be using <code>ts_mutate()</code> throughout.</strong>
</center>
</div>
</section>
<section id="lets-take-the-model-from-earlier" class="level2">
<h2 class="anchored" data-anchor-id="lets-take-the-model-from-earlier">Let‚Äôs take the <code>model</code> from earlier‚Ä¶</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="globe-slendr-2023-03-22_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="and-simulate-tree-sequence-from-it" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="and-simulate-tree-sequence-from-it">‚Ä¶ and simulate tree sequence from it</h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">10e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This simulates 1 Mb of sequence, with given rates of recombination and mutation.</p>
<p><br></p>

<div class="no-row-height column-margin column-container"><div class="">
<p>A tree sequence can be saved to disk with <code>ts_save()</code>.</p>
</div></div></section>
</section>
<section id="what-can-we-do-with-it" class="level1">
<h1>What can we do with it?</h1>
<section id="conversion-to-other-file-formats" class="level2">
<h2 class="anchored" data-anchor-id="conversion-to-other-file-formats">Conversion to other file formats</h2>
<p>If you have a tree-sequence object <code>ts</code>, you can do‚Ä¶</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_vcf</span>(ts, <span class="at">path =</span> <span class="st">"path/to/a/file.vcf.gz"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_eigenstrat</span>(ts, <span class="at">prefix =</span> <span class="st">"path/to/eigenstrat/prefix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">

</div>
<div class="cell" data-output-location="fragment">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_genotypes</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>1 multiallelic sites (0.076% out of 1320 total) detected and removed</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,319 √ó 5
     pos pop1_1_chr1 pop1_1_chr2 pop2_1_chr1 pop2_1_chr2
   &lt;int&gt;       &lt;int&gt;       &lt;int&gt;       &lt;int&gt;       &lt;int&gt;
 1 12391           0           0           1           0
 2 27991           0           0           1           1
 3 31249           0           0           1           1
 4 32284           0           0           1           0
 5 39497           1           1           0           0
 6 41476           1           1           0           0
 7 42022           0           0           1           1
 8 44384           0           0           1           1
 9 56882           0           0           1           1
10 69058           1           1           0           0
# ‚Ä¶ with 1,309 more rows</code></pre>
</div>
</div>
<div class="cell">

</div>
</section>
<section id="slendrs-r-interface-to-tskit" class="level2">
<h2 class="anchored" data-anchor-id="slendrs-r-interface-to-tskit"><em>slendr</em>‚Äôs R interface to <a href="https://tskit.dev/tskit"><em>tskit</em></a></h2>
<center>
<p><img src="images/slendr_tskit.png" class="img-fluid"></p>
<p>This <a href="https://www.slendr.net/reference/index.html#tree-sequence-loading-and-processing">R interface</a> links to Python methods implemented in <a href="https://tskit.dev/tskit/docs/stable/python-api.html#statistics"><em>tskit</em></a>.</p>
</center>
</section>
<section id="we-need-to-refer-to-samples" class="level2">
<h2 class="anchored" data-anchor-id="we-need-to-refer-to-samples">We need to refer to ‚Äúsamples‚Äù</h2>
<p>Each individual in <em>slendr</em> has a symbolic name and other metadata stored in the tree-sequence object.</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="globe-slendr-2023-03-22_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="extracting-sample-information" class="level2">
<h2 class="anchored" data-anchor-id="extracting-sample-information">Extracting sample information</h2>
<p>If we have a tree sequence <code>ts</code>, then we can get samples stored in it via <code>ts_samples()</code>:</p>
<div class="columns">
<div class="column" style="width:55%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,700 √ó 3
   name     time pop  
   &lt;chr&gt;   &lt;dbl&gt; &lt;chr&gt;
 1 pop1_1      0 pop1 
 2 pop1_2      0 pop1 
 3 pop1_3      0 pop1 
 4 pop1_4      0 pop1 
 5 pop1_5      0 pop1 
 6 pop1_6      0 pop1 
 7 pop1_7      0 pop1 
 8 pop1_8      0 pop1 
 9 pop1_9      0 pop1 
10 pop1_10     0 pop1 
# ‚Ä¶ with 4,690 more rows</code></pre>
</div>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">count</span>(pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 √ó 2
  pop       n
  &lt;chr&gt; &lt;int&gt;
1 pop1   1000
2 pop2    100
3 pop3   2500
4 pop4    700
5 pop5    400</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="popgen-statistics-in-slendr" class="level2">
<h2 class="anchored" data-anchor-id="popgen-statistics-in-slendr">Popgen statistics in <em>slendr</em></h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img src="images/slendr_ts_functions.png" class="img-fluid"></p>
</div><div class="column" style="width:50%;">
<p>These functions work with:</p>
<ul>
<li>vectors of names (single population)</li>
<li>lists of vectors of names (multiple populations)</li>
</ul>
<p>The first argument is <em>always</em> a <code>ts</code> object.</p>
</div>
</div>
<div class="incremental">
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a><span class="fu">ts_f3</span>(ts, <span class="at">A =</span> <span class="st">"popA_1"</span>, <span class="at">B =</span> <span class="fu">c</span>(<span class="st">"popB_1"</span>, <span class="st">"popB_2"</span>), <span class="at">C =</span> <span class="st">"outgroup_1"</span>)</span>
<span id="cb39-2"><a href="#cb39-2"></a><span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> <span class="fu">list</span>(<span class="at">A =</span> <span class="fu">c</span>(<span class="st">"popA_1"</span>, <span class="st">"popA_2"</span>, <span class="st">"popA_3"</span>),</span>
<span id="cb39-3"><a href="#cb39-3"></a>                                    <span class="at">B =</span> <span class="fu">c</span>(<span class="st">"popB_1"</span>, <span class="st">"popB_2"</span>, <span class="st">"popB_3"</span>),</span>
<span id="cb39-4"><a href="#cb39-4"></a>                                    <span class="at">C =</span> <span class="fu">c</span>(<span class="st">"popB_1"</span>, <span class="st">"popB_2"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
</section>
<section id="exercise-2-popgen-statistics" class="level1 page-columns page-full">
<h1>Exercise #2 ‚Äî popgen statistics</h1>
<section id="exercise-2-popgen-statistics-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-popgen-statistics-1">Exercise #2 ‚Äî popgen statistics</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/intro_model1.png" class="img-fluid figure-img" style="width:70.0%"></p>
</figure>
</div>
</section>
<section id="exercise-2-popgen-statistics-2" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-popgen-statistics-2">Exercise #2 ‚Äî popgen statistics</h2>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="globe-slendr-2023-03-22_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="exercise-2-popgen-statistics-3" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="exercise-2-popgen-statistics-3">Exercise #2 ‚Äî popgen statistics</h2>
<p>Simulate a tree sequence using <code>msprime()</code> from your model (perhaps 50Mb?), add mutations with <code>ts_mutate()</code>. Inspect its contents using <code>ts_samples()</code>.</p>
<hr>
<p>Compute <a href="https://en.wikipedia.org/wiki/Nucleotide_diversity">diversity (<span class="math inline">\(\pi\)</span>)</a> in each population (<code>ts_diversity()</code>). Do your results fit <span class="math inline">\(N_e\)</span> in your model?</p>
<hr>
<p>Compute <code>ts_divergence()</code> (genetic divergence) or <code>ts_fst()</code> (<a href="https://en.wikipedia.org/wiki/Fixation_index"><span class="math inline">\(F_{ST}\)</span></a>) between pairs of populations. Do the results match the order of splits in your model?</p>

<div class="no-row-height column-margin column-container"><div class="">
<p><em>Hint:</em> See <a href="https://www.slendr.net/reference/index.html#tree-sequence-statistics">manual page</a> of each tree-sequence statistic function for notes on how to use it.</p>
</div></div></section>
<section id="exercise-2-popgen-statistics-4" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-popgen-statistics-4">Exercise #2 ‚Äî popgen statistics</h2>
<p>An R trick for using names of samples in <code>ts_*()</code> functions:</p>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,700 √ó 3
   name     time pop  
   &lt;chr&gt;   &lt;dbl&gt; &lt;chr&gt;
 1 pop1_1      0 pop1 
 2 pop1_2      0 pop1 
 3 pop1_3      0 pop1 
 4 pop1_4      0 pop1 
 5 pop1_5      0 pop1 
 6 pop1_6      0 pop1 
 7 pop1_7      0 pop1 
 8 pop1_8      0 pop1 
 9 pop1_9      0 pop1 
10 pop1_10     0 pop1 
# ‚Ä¶ with 4,690 more rows</code></pre>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get a list of lists of individual names</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">split</span>(., .<span class="sc">$</span>pop) <span class="sc">%&gt;%</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(pull, <span class="st">"name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 5
 $ pop1: chr [1:1000] "pop1_1" "pop1_2" "pop1_3" "pop1_4" ...
 $ pop2: chr [1:100] "pop2_1" "pop2_2" "pop2_3" "pop2_4" ...
 $ pop3: chr [1:2500] "pop3_1" "pop3_2" "pop3_3" "pop3_4" ...
 $ pop4: chr [1:700] "pop4_1" "pop4_2" "pop4_3" "pop4_4" ...
 $ pop5: chr [1:400] "pop5_1" "pop5_2" "pop5_3" "pop5_4" ...</code></pre>
</div>
</div>
</div>
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(samples<span class="sc">$</span>pop1, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop1_1" "pop1_2" "pop1_3"</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-2-solution" class="level1">
<h1>Exercise #2 ‚Äî solution</h1>
<p>Solution is on GitHub at <a href="https://github.com/bodkan/globe-slendr-2023-03-22/tree/main/solutions/ex2.R">this</a> link.</p>
</section>
<section id="gene-flow-admixture" class="level1">
<h1>Gene flow / admixture</h1>
<section id="gene_flow-events" class="level2">
<h2 class="anchored" data-anchor-id="gene_flow-events"><code>gene_flow()</code> events</h2>
<p>Gene flow is programmed the <code>gene_flow()</code> function.</p>
<p><br></p>
<p>If we have <code>p1</code> and <code>p2</code>, we schedule gene flow with:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> p1, <span class="at">to =</span> p2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p>Multiple gene-flow events can be gathered in a list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> p1, <span class="at">to =</span> p2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>),</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> <span class="sc">&lt;</span>..<span class="sc">&gt;</span>, <span class="at">to =</span> <span class="sc">&lt;</span>..<span class="sc">&gt;</span>, <span class="at">start =</span> <span class="sc">&lt;</span>...<span class="sc">&gt;</span>, <span class="at">end =</span> <span class="sc">&lt;</span>...<span class="sc">&gt;</span>, <span class="at">rate =</span> <span class="sc">&lt;</span>...<span class="sc">&gt;</span>),</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="sc">&lt;</span> potentially many more ... <span class="sc">&gt;</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="incremental">
<p><br></p>
<center>
<strong><code>gene_flow()</code> checks for consistency!</strong>
</center>
</div>
</section>
<section id="adding-gene-flows-to-the-model" class="level2">
<h2 class="anchored" data-anchor-id="adding-gene-flows-to-the-model">Adding gene flow(s) to the model</h2>
<p><br> <br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1"></a><span class="fu">library</span>(slendr)</span>
<span id="cb50-2"><a href="#cb50-2"></a><span class="fu">init_env</span>()</span>
<span id="cb50-3"><a href="#cb50-3"></a></span>
<span id="cb50-4"><a href="#cb50-4"></a><span class="sc">&lt;</span>... your definitions of populations ...<span class="sc">&gt;</span></span>
<span id="cb50-5"><a href="#cb50-5"></a>  </span>
<span id="cb50-6"><a href="#cb50-6"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> ..., <span class="at">to =</span> ..., <span class="sc">&lt;</span>start, end, rate<span class="sc">&gt;</span>)</span>
<span id="cb50-7"><a href="#cb50-7"></a></span>
<span id="cb50-8"><a href="#cb50-8"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb50-9"><a href="#cb50-9"></a>  <span class="at">populations =</span> <span class="fu">list</span>(...),</span>
<span id="cb50-10"><a href="#cb50-10"></a>  <span class="at">gene_flow =</span> gf,           <span class="co"># &lt;--- gene flow specified above</span></span>
<span id="cb50-11"><a href="#cb50-11"></a>  <span class="at">generation_time =</span> <span class="dv">30</span></span>
<span id="cb50-12"><a href="#cb50-12"></a>)</span>
<span id="cb50-13"><a href="#cb50-13"></a></span>
<span id="cb50-14"><a href="#cb50-14"></a>ts <span class="ot">&lt;-</span> <span class="er">&lt;</span>... simulating a tree sequence ....<span class="sc">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="exercise-3" class="level1 page-columns page-full">
<h1>Exercise #3</h1>
<section id="exercise-3a-adding-gene_flow" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3a-adding-gene_flow">Exercise #3a ‚Äî adding <code>gene_flow()</code></h2>
<p>Add introgression into your Exercise #1 model (5% pulse NEA -&gt; EUR over 55-50 kya) and simulate a tree sequence from it.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/intro_model1.png" class="img-fluid figure-img" style="width:70.0%"></p>
</figure>
</div>
</section>
<section id="exercise-3a-adding-gene_flow-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3a-adding-gene_flow-1">Exercise #3a ‚Äî adding <code>gene_flow()</code></h2>
<p>Add introgression into your Exercise #1 model (5% pulse NEA -&gt; EUR over 55-50 kya) and simulate a tree sequence from it.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/intro_model2.png" class="img-fluid figure-img" style="width:70.0%"></p>
</figure>
</div>
</section>
<section id="exercise-3b-detect-introgression" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="exercise-3b-detect-introgression">Exercise #3b ‚Äî detect introgression</h2>
<p>Pretend that you don‚Äôt know whether or not you simulated Neanderthal introgression in your model.</p>
<hr>
<p>Test for the evidence of Neanderthal ancestry in Europeans vs Africans using the <span class="math inline">\(f_4\)</span>-statistic implemented <code>ts_f4()</code>.</p>

<div class="no-row-height column-margin column-container"><div class="">
<p><em>Hint</em>: <span class="math inline">\(f_4\)</span> (W = AFR, X = EUR; Y = NEAND, Z = CHIMP) is the test setup where the null hypothesis states that the value of this statistic is ‚Äúconsistent with zero‚Äù. ‚ÄúSignificantly negative value‚Äù indicates gene flow NEAND-EUR.</p>
<p>You can run the test on single individuals from each population but the signal will be clearer if you compute the <span class="math inline">\(f_4\)</span> on ~20 Africans and Europeans.</p>
</div></div></section>
</section>
<section id="exercise-3-solution" class="level1">
<h1>Exercise #3 ‚Äî solution</h1>
<p>Solution is on GitHub at <a href="https://github.com/bodkan/globe-slendr-2023-03-22/tree/main/solutions/ex3.R">this</a> link.</p>
</section>
<section id="time-series-data" class="level1">
<h1>Time-series data</h1>
<section id="sampling-adna-samples-through-time" class="level2">
<h2 class="anchored" data-anchor-id="sampling-adna-samples-through-time">Sampling aDNA samples through time</h2>
<p><br></p>
<p>Imagine we have <code>pop1</code>, <code>pop2</code>, ‚Ä¶ compiled in a <code>model</code>.</p>
<p><br></p>
<div class="incremental">
<p>To record <em>ancient</em> individuals in the tree sequence, we can use the function <code>schedule_sampling()</code>:</p>
</div>
<div class="incremental">
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  model,                <span class="co"># compiled slendr model object</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">500</span>),  <span class="co"># at these times (can be also a single number) ...</span></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, <span class="dv">42</span>),       <span class="co"># ... sample 42 individuals from pop1</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop2, <span class="dv">10</span>),       <span class="co"># ... sample 10 individuals from pop2</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop3, <span class="dv">1</span>)         <span class="co"># ... sample 1 individual from pop 3</span></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="sampling-schedule-format" class="level2">
<h2 class="anchored" data-anchor-id="sampling-schedule-format">Sampling schedule format</h2>
<p>The output of <code>schedule_sampling()</code> is a plain data frame:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 √ó 7
   time pop       n y_orig x_orig y     x    
  &lt;int&gt; &lt;chr&gt; &lt;int&gt; &lt;lgl&gt;  &lt;lgl&gt;  &lt;lgl&gt; &lt;lgl&gt;
1 10000 EUR       1 NA     NA     NA    NA   
2 20000 EUR       1 NA     NA     NA    NA   
3 30000 EUR       1 NA     NA     NA    NA   
4 40000 EUR       1 NA     NA     NA    NA   </code></pre>
</div>
</div>
<div class="incremental">
<p>We can bind multiple sampling schedules together, giving us finer control about sampling:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>eur_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">40000</span>, <span class="dv">30000</span>, <span class="dv">20000</span>, <span class="dv">10000</span>, <span class="dv">0</span>), <span class="fu">list</span>(eur, <span class="dv">1</span>))</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>afr_samples <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(model, <span class="at">times =</span> <span class="dv">0</span>, <span class="fu">list</span>(afr, <span class="dv">1</span>))</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">rbind</span>(eur_samples, afr_samples)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="how-to-use-a-sampling-schedule" class="level2">
<h2 class="anchored" data-anchor-id="how-to-use-a-sampling-schedule">How to use a sampling schedule?</h2>
<p>To sample individuals based on a given schedule, we use the <code>samples =</code> argument of the <code>msprime()</code> functions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">1e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p>We can verify that only specific individuals are recorded:</p>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 √ó 3
  name   time pop  
  &lt;chr&gt; &lt;dbl&gt; &lt;chr&gt;
1 EUR_1 40000 EUR  
2 EUR_2 30000 EUR  
3 EUR_3 20000 EUR  
4 EUR_4 10000 EUR  
5 AFR_1     0 AFR  
6 EUR_5     0 EUR  </code></pre>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>‚ïî‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïó
‚ïëTreeSequence               ‚ïë
‚ï†‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï§‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï£
‚ïëTrees          ‚îÇ       1343‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëSequence Length‚îÇ    1000000‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëTime Units     ‚îÇgenerations‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëSample Nodes   ‚îÇ         12‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëTotal Size     ‚îÇ  184.5 KiB‚ïë
‚ïö‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïß‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïù
‚ïî‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï§‚ïê‚ïê‚ïê‚ïê‚ï§‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï§‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïó
‚ïëTable      ‚îÇRows‚îÇSize     ‚îÇHas Metadata‚ïë
‚ï†‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï£
‚ïëEdges      ‚îÇ3982‚îÇ124.4 KiB‚îÇ          No‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëIndividuals‚îÇ   6‚îÇ192 Bytes‚îÇ          No‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëMigrations ‚îÇ   0‚îÇ  8 Bytes‚îÇ          No‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëMutations  ‚îÇ   0‚îÇ 16 Bytes‚îÇ          No‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëNodes      ‚îÇ 924‚îÇ 25.3 KiB‚îÇ          No‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëPopulations‚îÇ   4‚îÇ341 Bytes‚îÇ         Yes‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëProvenances‚îÇ   1‚îÇ  2.6 KiB‚îÇ          No‚ïë
‚ïü‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îº‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚ï¢
‚ïëSites      ‚îÇ   0‚îÇ 16 Bytes‚îÇ          No‚ïë
‚ïö‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïß‚ïê‚ïê‚ïê‚ïê‚ïß‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïß‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïù</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-4" class="level1 page-columns page-full">
<h1>Exercise #4</h1>
<section id="exercise-4a-ancient-samples" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4a-ancient-samples">Exercise #4a ‚Äî ancient samples</h2>
<div class="cell">
<div class="cell-output-display">
<p><img src="globe-slendr-2023-03-22_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="exercise-4a-ancient-samples-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4a-ancient-samples-1">Exercise #4a ‚Äî ancient samples</h2>
<p>Simulate data from your <code>model</code> using this sampling:</p>
<ul>
<li>one present-day CHIMP</li>
<li>20 present-day EUR and AFR individuals</li>
<li>1 NEA at 70 ky, 1 NEA at 40 ky</li>
<li>1 EUR every 1000 years between 50-5 kya</li>
</ul>
<p><em>Hint</em>: you can do this by:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="co"># rbind(...) together individual schedule_sampling() data frames</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">samples =</span> samples, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="exercise-4b-f_4-ratio-statistic" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="exercise-4b-f_4-ratio-statistic">Exercise #4b ‚Äî <span class="math inline">\(f_4\)</span>-ratio statistic</h2>
<p>Use <span class="math inline">\(f_4\)</span>-ratio statistic to replicate the <a href="https://www.pnas.org/doi/10.1073/pnas.1814338116#fig01">following figure</a>:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/neand_decline.png" class="img-fluid figure-img" style="width:60.0%"></p>
</figure>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p><em>Hint</em>: You can compute Neanderthal ancestry for a vector of individual names <code>X</code> as <code>ts_f4ratio(ts, X = X, "NEA_1", "NEA_2", "AFR_1", "CHIMP_1")</code> (‚Äúdirect f4-ratio‚Äù from <a href="https://www.pnas.org/doi/full/10.1073/pnas.1814338116">Petr <em>et al</em>., PNAS 2019</a>).</p>
</div></div></section>
</section>
<section id="exercise-4-solution" class="level1">
<h1>Exercise #4 ‚Äî solution</h1>
<p>Solution is on GitHub at <a href="https://github.com/bodkan/globe-slendr-2023-03-22/tree/main/solutions/ex4.R">this</a> link.</p>
<section id="more-information" class="level2">
<h2 class="anchored" data-anchor-id="more-information">More information</h2>
<p><br></p>
<ul>
<li><p><em>slendr</em> paper on <a href="https://www.biorxiv.org/content/10.1101/2022.03.20.485041v3">bioRxiv</a></p></li>
<li><p>documentation tutorials can be found <a href="https://www.slendr.net">here</a></p></li>
<li><p>GitHub repo (bug reports!) is <a href="http://github.com/bodkan/slendr">here</a></p></li>
</ul>
<hr>
<ul>
<li>new R package for ABC modeling <a href="http://github.com/bodkan/demografr"><em>demografr</em></a></li>
</ul>
<hr>
<br>
<center>
<strong>For questions, reach out at <u>mp@bodkan.net</u>!</strong>
</center>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>